network/p2p/p2p.go

```diff
@@ -256,73 +255,174 @@ func (n *p2pNetwork) Start(logger *zap.Logger) error {
+				continue
+			}
+			n.libConnManager.Unprotect(peer, peers.ProtectedTag)
+		}
+		connMgr.TrimPeers(ctx, logger, n.host.Network(), maxPeersToDrop) // trim up to maxPeersToDrop
+	}
+}
 
-		// Disconnect from irrelevant peers
-		disconnectedPeers := connMgr.DisconnectFromIrrelevantPeers(logger, maximumIrrelevantPeersToDisconnect, n.host.Network(), allPeers, mySubnets)
-		if disconnectedPeers > 0 {
-			return
+// PeerProtection returns a map of protected peers based on the intersection of subnets.
+// it protects the best peers by these rules:
+// - At least 2 peers per subnet.
+// - Prefer peers that you have more shared subents with.
+// - Protect at most immunityQuota peers.
+func (n *p2pNetwork) PeerProtection(
+	immunityQuota int,
+) map[peer.ID]struct{} {
+	protectedPeers := make(map[peer.ID]struct{})
+
+	tpcs := n.topicsCtrl.Topics()
+	peerz := make(map[string][]peer.ID, len(tpcs)) // topic -> peers
+	for _, tpc := range tpcs {
+		var err error
+		peerz[tpc], err = n.topicsCtrl.Peers(tpc)
+		if err != nil {
+			n.interfaceLogger.Error(
+				"Cant get peers for topic, skipping to keep the network running",
+				zap.String("topic", tpc),
+				zap.Error(err),
+			)
+			continue
 		}
+	}
 
-		// Trim peers according to subnet participation (considering the subnet size)
-		connMgr.TagBestPeers(logger, n.cfg.MaxPeers-1, mySubnets, allPeers, n.cfg.TopicMaxPeers)
-		connMgr.TrimPeers(ctx, logger, n.host.Network())
+	sharedTopics := func(peerID peer.ID) int {
+		shared := 0
+		for _, tpc := range tpcs {
+			if slices.Contains(peerz[tpc], peerID) {
+				shared++
+			}
+		}
+		return shared
 	}
+
+	// for each topic we have peers in sort these peers by the number of topics they share with us
+	// and protect top 2 peers for each topic, if there's only one peer in that topic, always protect it
+	for _, tpc := range tpcs {
+		peersInTopic := peerz[tpc]
+		if len(peersInTopic) == 0 {
+			continue
+		}
+		if len(peersInTopic) == 1 {
+			protectedPeers[peersInTopic[0]] = struct{}{}
+			immunityQuota--
+			continue
+		}
+
+		slices.SortFunc(peersInTopic, func(a, b peer.ID) int {
+			x := sharedTopics(a)
+			y := sharedTopics(b)
+			return x - y // desc order
+		})
+
+		const minPeersPerTopic = 2
+		for i := 0; i < min(minPeersPerTopic, len(peersInTopic)) && immunityQuota > 0; i++ {
+			protectedPeers[peersInTopic[i]] = struct{}{}
+			immunityQuota--
+		}
+	}
+
+	return protectedPeers
 }
 
 // startDiscovery starts the required services
 // it will try to bootstrap discovery service, and inject a connect function.
 // the connect function checks if we can connect to the given peer and if so passing it to the backoff connector.
 func (n *p2pNetwork) startDiscovery(logger *zap.Logger, connector chan peer.AddrInfo) {
-	discoveredPeers := make(chan peer.AddrInfo, connectorQueueSize)
-	go func() {
-		ctx, cancel := context.WithCancel(n.ctx)
-		defer cancel()
-		n.backoffConnector.Connect(ctx, discoveredPeers)
```

Comment:
I don't understand how the connector works without this?

Reply:
Oh yeah, I should have left a comment for this change here, 

it's just dead code - what we really use is the argument we get in this function `connector chan peer.AddrInfo`